File: tools/csharp-roslyn-host/Rules/CodeQuality/PartialElementMissingAccessModifier.cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;

namespace TrueCourse.RoslynHost;

/// <summary>
/// A `partial` type (or partial method) declaration that omits an explicit access
/// modifier. Because the accessibility may be stated in a *different* part of the
/// partial, this declaration on its own reads as having unknown accessibility. We flag
/// the partial declaration that carries no access modifier — even if another part
/// does — so each part is self-documenting. SA1205.
/// </summary>
internal sealed class PartialElementMissingAccessModifier : ISemanticRule
{
    public string RuleKey => "code-quality/deterministic/partial-element-missing-access-modifier";

    private static readonly string[] AccessModifiers = { "public", "private", "protected", "internal" };

    public IEnumerable<Violation> Analyze(SemanticModel model, SyntaxTree tree)
    {
        foreach (var node in tree.GetRoot().DescendantNodes())
        {
            if (node is not MemberDeclarationSyntax member) continue;
            if (!member.Modifiers.Any(m => m.IsKind(SyntaxKind.PartialKeyword))) continue;
            // Nested partial types inside an interface inherit implicit public; skip.
            if (member.Parent is InterfaceDeclarationSyntax) continue;
            if (HasExplicitAccessModifier(member.Modifiers)) continue;

            var (loc, label) = Target(member);
            if (loc is null) continue;
            var pos = loc.GetLineSpan().StartLinePosition;
            yield return new Violation(
                RuleKey, tree.FilePath, pos.Line + 1, pos.Character + 1,
                $"Partial {label} declares no access modifier here, hiding its accessibility in another part; state it explicitly on every part.");
        }
    }

    private static bool HasExplicitAccessModifier(SyntaxTokenList modifiers) =>
        modifiers.Any(m => AccessModifiers.Contains(m.Text));

    private static (Location?, string) Target(MemberDeclarationSyntax member) => member switch
    {
        BaseTypeDeclarationSyntax t => (t.Identifier.GetLocation(), $"type '{t.Identifier.ValueText}'"),
        MethodDeclarationSyntax m when m.ExplicitInterfaceSpecifier is null
            => (m.Identifier.GetLocation(), $"method '{m.Identifier.ValueText}'"),
        _ => (null, ""),
    };
}

File: packages/analyzer/src/rules/code-quality/visitors/csharp/partial-element-missing-access-modifier.ts
import type { Node as SyntaxNode } from 'web-tree-sitter'
import type { CodeRuleVisitor } from '../../../types.js'
import { makeViolation } from '../../../types.js'

const ACCESS_MODIFIERS = new Set(['public', 'private', 'protected', 'internal'])

const PARTIAL_TYPES = [
  'class_declaration', 'struct_declaration', 'interface_declaration', 'record_declaration',
  'method_declaration',
]

function modifierTexts(node: SyntaxNode): string[] {
  return node.children.filter((c) => c?.type === 'modifier').map((c) => c!.text)
}

/**
 * A `partial` type or method part that states no access modifier on this part. The
 * accessibility is then declared only on another part (or defaulted), so a reader
 * of this file cannot tell how visible the element is. State the modifier on every
 * part. Interface members (implicitly public) and explicit interface
 * implementations are excluded.
 */
export const csharpPartialElementMissingAccessModifierVisitor: CodeRuleVisitor = {
  ruleKey: 'code-quality/deterministic/partial-element-missing-access-modifier',
  languages: ['csharp'],
  nodeTypes: PARTIAL_TYPES,
  visit(node, filePath, sourceCode) {
    const mods = modifierTexts(node)
    if (!mods.includes('partial')) return null
    if (mods.some((m) => ACCESS_MODIFIERS.has(m))) return null
    if (node.parent?.parent?.type === 'interface_declaration') return null

    if (node.type === 'method_declaration') {
      const nm = node.childForFieldName('name')
      if (nm?.type === 'qualified_name' || (nm?.text ?? '').includes('.')) return null
    }

    const target = node.childForFieldName('name') ?? node
    return makeViolation(
      this.ruleKey, target, filePath, 'low',
      'Partial element missing access modifier',
      'Partial declaration states no access modifier here, hiding its accessibility in another part; state it explicitly on every part.',
      sourceCode,
      'Add the access modifier to every part of the partial declaration.',
    )
  },
}
